### PR TITLE
Create Entry for cluster

### PR DESCRIPTION
**Pull request by bonifacyj**
This is a next step as described in the Implementation Strategy of the [design doc](https://docs.google.com/document/d/1v8E5WjGhalYGUNj9erq8LiGFhlOtU62I9k9vgM_ivVY/edit?usp=sharing).

Code implements creation of DataCatalog entries for a Kafka cluster. Tests are included.

### CODE REVIEW

File: google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/assembled_entry_factory_test.py [hunk 1]
```diff
@@ -13,10 +15,7 @@ class AssembledEntryFactoryTestCase(unittest.TestCase):
     __LOCATION_ID = 'location_id'
```
Comment (socha23) [inline google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/assembled_entry_factory_test.py:15]: Do we even need those constants if they are used only once, in the same file?

File: google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/datacatalog_entry_factory_test.py [hunk 2]
```diff
@@ -45,3 +44,20 @@ def test_topic_metadata_should_be_converted_to_datacatalog_entries(
...
+
+    def test_cluster_metadata_should_be_converted_to_dc_entries(
+            self, entry_path):
+        entry_path.return_value = \
+            DataCatalogEntryFactoryTestCase.__MOCKED_ENTRY_PATH
+        metadata = utils.Utils.convert_json_to_object(self.__MODULE_PATH,
+                                                      'test_metadata.json')
+        cluster_id = metadata[MetadataDictKeys.CLUSTER_ID]
+        entry_id, entry = self.__entry_factory.make_entry_for_cluster(metadata)
+        self.assertIsNotNone(entry_id)
```
Comment (socha23) [inline google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/datacatalog_entry_factory_test.py:52]: can we compare whole entry to what we expect with one assertion?

File: google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/test_utils/test_utils.py [hunk 3]
```diff
@@ -14,16 +12,26 @@ def make_entry_for_topic(self, topic):
         return entry_id, entry
 
 
-class FakeMetadataScraper(MetadataScraper):
+class FakeKafkaConsumer(mock.MagicMock):
 
-    def _get_metadata_from_message_broker_connection(self, connection_args):
-        return mock.MagicMock()
+    def list_topics(self, timeout=-1):
+        raw_metadata = mock.MagicMock()
```
Comment (socha23) [inline google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/test_utils/test_utils.py]: so is there any reason to use MagicMock instead of some hardcoded dicts here?

File: google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/assembled_entry_factory_test.py (hunk 1 above)
Comment (bonifacyj) [reply google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/assembled_entry_factory_test.py:15]: We will probably need them later again, because there are more tests coming to this file, when I add tags creation

File: google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/test_utils/test_utils.py (hunk 3 above)
Comment (bonifacyj) [reply google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/test_utils/test_utils.py]: to convey the idea that original list_topics() returns a python object of a ClusterMetadata class, that is potentially more than a dict. At the same time, from functionality perspective, we can just do hardcoded dict. Do you find it will be easier to read that way?

File: google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/datacatalog_entry_factory_test.py (hunk 2 above)
Comment (bonifacyj) [reply google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/datacatalog_entry_factory_test.py:52]: done :)

File: google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/assembled_entry_factory_test.py (hunk 1 above)
Comment (socha23) [reply google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/assembled_entry_factory_test.py:15]: okay, makes sense

File: google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/test_utils/test_utils.py (hunk 3 above)
Comment (socha23) [reply google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/test_utils/test_utils.py]: in general I'm against using mocks if they aren't needed. If the 'real' version of this object returns a ClusterMetadata object, can we do the same here (perhaps passing the hardcoded values to the ClusterMetadata constructor?

Comment (bonifacyj) [reply google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/test_utils/test_utils.py]: Hm, it is possible, I think, just not through the constructor, but by setting the instance variables, as [source suggests](https://docs.confluent.io/current/clients/confluent-kafka-python/_modules/confluent_kafka/admin.html#ClusterMetadata). I'll try to refactor it then.

File: google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/datacatalog_entry_factory_test.py [hunk 4]
```diff
@@ -22,42 +25,51 @@ def setUp(self):
...
-        self.assertEqual('kafka', entry.user_specified_system)
-        self.assertEqual(DataCatalogEntryFactoryTestCase.__MOCKED_ENTRY_PATH,
-                         entry.name)
-        self.assertIn(DataCatalogEntryFactoryTestCase.__METADATA_SERVER_HOST,
-                      entry.linked_resource)
-        self.assertIn(cluster_id, entry.display_name)
+        self.assertEqual(entry, expected_entry)
+
+    def compare_entries(self, entry1, entry2, msg=None):
+        if not msg:
```
Comment (socha23) [inline google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/datacatalog_entry_factory_test.py]: I wonder if there is a way to not compare field-by-field by hand. Perhaps comparing string representations of Entries would be enough?

**Review by socha23 — APPROVED** [review]

File: google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/test_utils/test_utils.py (hunk 3 above)
Comment (bonifacyj) [reply google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/test_utils/test_utils.py]: Done

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/config/metadata_constants.py [hunk 5]
```diff
@@ -0,0 +1,4 @@
+class MetadataDictKeys:
```
Comment (mesmacosta) [inline google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/config/metadata_constants.py]: Class name should follow the convention of the file name: `metadata_constants.py.`

Rename to MetadataConstants ?

Otherwise could rename the file.

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/datacatalog_cli.py [hunk 6]
```diff
@@ -86,8 +88,10 @@ def _parse_args(self, argv):
                             help='Enables monitoring metrics on the connector')
         return parser.parse_args(argv)
 
-    def _get_connection_args(self, args):
-        return {'bootstrap.servers': args.kafka_host, 'group.id': 'kafka2dc'}
+    def _get_kafka_consumer(self, args):
```
Comment (mesmacosta) [inline google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/datacatalog_cli.py]: Any reason why the CLI class should create the Kafka Consumer?

From a code design point of view, the `consumer` is only required for the `scraper` class.

This is adding an implementation detail to the CLI code that's not needed, if there is a good reason for the CLI to create it, discuss this.

Otherwise I think this should be done by the Scraper class WDYT?

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_scraper.py [hunk 7]
```diff
@@ -1,33 +1,37 @@
...
+from google.datacatalog_connectors.kafka.config.\
+    metadata_constants import MetadataDictKeys
 
 
 class MetadataScraper:
 
-    def __init__(self):
-        pass
+    def __init__(self, consumer):
+        self._consumer = consumer
```
Comment (mesmacosta) [inline google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_scraper.py:9]: Same as the CLI class comment.

**Review by mesmacosta — CHANGES_REQUESTED** [review]

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/datacatalog_cli.py (hunk 6 above)
Comment (bonifacyj) [reply google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/datacatalog_cli.py]: The idea to move Consumer creation to the CLI code appeared when I had problem testing get_metadata() method. I used to have Consumer creation in this method, ([see line 14-15 here](https://github.com/bonifacyj/datacatalog-connectors-message-brokers/pull/2/commits/c59e530edbf028004e7be02b79c1d3e74568eba1#diff-0c0cf672ecfdd1c2488e39baf694c2f0)), which made it so hard to test. Then I had the following two choices:
1. Pull consumer creation into another method within the MetadataScraper and then in the test mock the return value of that method (returning a FakeConsumer)
or
2. Pull consumer creation up outside from the MetadataScraper class and give it the Consumer as an argument in the constructor. 

We discussed both possibilities with @socha23 and he pointed out that he feels like scraping metadata process should not include Consumer creation, and he would expect that MetadataScraper already receives the Consumer as an input. 

At the same time, I see your point. It also would be consistent with all the other connectors to establish the connection to data source in the MetadataScraper class. As for the testing, I would then just go with option 1 from described above.

 @mesmacosta, @socha23, what do you think?

File: google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/datacatalog_entry_factory_test.py (hunk 4 above)
Comment (bonifacyj) [reply google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/datacatalog_entry_factory_test.py]: Done! Thanks for your suggestion, I didn't know it was an option.

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/datacatalog_cli.py (hunk 6 above)
Comment (mesmacosta) [reply google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/datacatalog_cli.py]: @bonifacyj I like the option 1.

a **con** of the CLI creating the consumer, is that users are able to use the connectors as client libraries, they don't have necessarily to run it as a CLI.

They can import the synchronizer directly on their own scripts. If the CLI class creates the Consumer it's harder to use it that way.

Python is really flexible regarding mocking, for example the test for oracle scraper class:
[metadata_scraper_test.py](https://github.com/GoogleCloudPlatform/datacatalog-connectors-rdbms/blob/master/google-datacatalog-oracle-connector/tests/google/datacatalog_connectors/oracle/scrape/metadata_scraper_test.py)

Also about the **Fake** test classes, they should be used only to test abstract classes, in the kafka connectors we don't have abstract classes, so we should not use it that way.

Let me go through some examples:

RDBMS:
[datacatalog_cli.py](https://github.com/GoogleCloudPlatform/datacatalog-connectors-rdbms/blob/master/google-datacatalog-rdbms-connector/src/google/datacatalog_connectors/rdbms/datacatalog_cli.py) Uses an Abstract class, so we need to create a FakeCLI class to implement it and test it.

HIVE:
[datacatalog_cli.py](https://github.com/GoogleCloudPlatform/datacatalog-connectors-hive/blob/master/google-datacatalog-hive-connector/src/google/datacatalog_connectors/hive/datacatalog_cli.py) is not an abstract class, so we have to test the real class: [datacatalog_cli_test.py](https://github.com/GoogleCloudPlatform/datacatalog-connectors-hive/blob/master/google-datacatalog-hive-connector/tests/google/datacatalog_connectors/hive/datacatalog_cli_test.py)

So to sum up:
First I'd suggest going for option 1 so we are consistent with all the other connectors.
Second the [datacatalog_cli_test.py#L31](https://github.com/bonifacyj/datacatalog-connectors-message-brokers/blob/entry-for-cluster/google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/datacatalog_cli_test.py#L31) should use the real class and not a fake one, like the example.

WDYT @socha23

**Review by mesmacosta — CHANGES_REQUESTED** [review]

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/datacatalog_cli.py (hunk 6 above)
Comment (socha23) [reply google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/datacatalog_cli.py]: If the synchronizer is also meant to be used directly, and not only through CLI, then I agree, it should create its dependencies. I thought about general Dependency injection rule from SOLID - getting collaborators from outside is preferable to creating them, but in this case turns out synchronizer IS the outside for some use cases. Yet another option would be to have synchronizer's constructor accept a consumer, but with a default value set to 'real' consumer so it's easy to use with default and yet possible to change in tests. 

If mocking methods is what other connectors use, let's do so so here too. 

I agree that testing real CLI class with changing only the KafkaConsumer is preferable to creating a fake CLI. Fakes are to be used with dependencies - that is, objects that the class under test calls; but not class under the test itself, we want to test the real class, not the fake.

Comment (mesmacosta) [reply google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/datacatalog_cli.py]: I agree about the scraper class receiving the consumer class in the constructor, and - _getting collaborators from outside is preferable to creating them_ -  makes it much easier to test.

Is there any benefit on also having the synchronizer's constructor accept the consumer? Because the Synchronizer class job is to orchestrate the connector execution, so having the consumer class created by it would make sense, but it doesn't necessarily need to be at the synchronizer's constructor.

Transferring the dependency creation to the **most** outside level, not sure if it's worth it. I mean we will have to mock the dependency at some level on testing. 

Ideally with dependency injection, you would have some configuration layer that sets up the dependencies. Python don't have good IoC frameworks, so if we have the dependencies at the constructor level for the scraper class, we already get the benefit of increasing testability.

If we expose it to another level, and have it outside of the synchronizer I feel that we are leaking implementation details, the CLI should be responsible only for parsing the CLI args and calling the synchronizer class. Also layers 
outside of the synchronizer class, should not be aware of this consumer class, would users need to create the consumer in any scenario? 

So, a suggestion would be to have the scraper class receive the consumer class in the constructor and have the synchronizer class create it before calling the scraper method.

The dependency creation would be done by the synchronizer class that already orchestrates the execution of the connector. And replace the fake CLI class with the real class on the test class.

Does it make sense @socha23 @bonifacyj ?

Comment (bonifacyj) [reply google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/datacatalog_cli.py]: Regarding the place for dependency injection - I think creating Consumer in the Synchronizer class is a great compromise between two approaches that we had here at the beginning (creating it in the Scraper vs by the CLI).  

Regarding the FakeCLI - the "fake" in the name is misleading, as I see now. In fact, it inherits from the real CLI and the only "fake" part in it is args parsing, which should be mocked anyways. So at the end of the day, the real CLI class ends up being tested.

I will make it more clear though by moving the mocked args parsing into the test itself, as done in @mesmacosta 's [example](https://github.com/GoogleCloudPlatform/datacatalog-connectors-hive/blob/master/google-datacatalog-hive-connector/tests/google/datacatalog_connectors/hive/datacatalog_cli_test.py).

Comment (socha23) [reply google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/datacatalog_cli.py]: Sounds good to me!

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/sync/datacatalog_synchronizer.py [hunk 8]
```diff
@@ -62,6 +62,14 @@ def run(self):
 
         return self.__task_id
 
+    def _create_consumer(self):
+        connection_config = {
+            'bootstrap.servers': self.__kafka_host,
+            'group.id': 'kafka2dc'
```
Comment (mesmacosta) [inline google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/sync/datacatalog_synchronizer.py]: @bonifacyj the `group.id` parameter should be in users' control, no?

they might use some pattern to name the Kafka consumer group in their production environment.  Because when  you connect to a topic with a consumer group, Kafka or Zookeper stores this information internally.

This should be at least a parameter in the constructor of the synchronizer class, and could have the default value `kafka2dc`.

**Review by mesmacosta — CHANGES_REQUESTED** [review]
@bonifacyj see the comment about the `group.id` WDYT?

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/sync/datacatalog_synchronizer.py (hunk 8 above)
Comment (bonifacyj) [reply google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/sync/datacatalog_synchronizer.py]: sure! I've made changes

**Review by mesmacosta — APPROVED** [review]
